lmdeploy/pytorch/models/utils/cudagraph.py

```diff
@@ -0,0 +1,171 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from dataclasses import dataclass
+from typing import Any, Dict, List
+
+import torch
+from torch import Tensor
+
+from lmdeploy.pytorch.model_inputs import StepContext
+
+BuffType = Dict[str, Tensor]
+
+
+def next_power_of_2(n: int):
+    """Return the smallest power of 2 greater than or equal to n."""
+    n -= 1
+    n |= n >> 1
+    n |= n >> 2
+    n |= n >> 4
+    n |= n >> 8
+    n |= n >> 16
+    n |= n >> 32
+    n += 1
+    return n
+
+
+@dataclass
+class CudaGraphMeta:
+    """Meta info of cudagraph."""
+    max_batchs: int
+    max_tokens: int
+    num_blocks: int
+    is_decoding: int
+    device: torch.device
+    input_buffers: BuffType = None
+    output_buffers: BuffType = None
+
+
+class CudaGraphMixin:
+    """mixin class to support cudagraph."""
+
+    def support_cuda_graph(
+        self,
+        input_ids: torch.Tensor,
+        position_ids: torch.Tensor,
+        past_key_values: List[List[torch.Tensor]],
+        attn_metadata: Any = None,
+        inputs_embeds: torch.Tensor = None,
+        **kwargs,
+    ):
+        """return True is model support cudagraph."""
+        seq_lens = input_ids.size(1)
+        return seq_lens <= 256
```

Comment:
What is the motivation here?

Reply:
graph is captured at pow2 seqlens, input would be padded to ensure static input size.
inputs with large seqlens might have a large amount of unnecessary padding.